cpp/src/arrow/compute/kernels/scalar_arithmetic.cc

```diff
@@ -355,6 +355,32 @@ struct PowerChecked {
   }
 };
 
+struct Round {
+  template <typename T, typename Arg>
+  static constexpr enable_if_floating_point<T> Call(KernelContext*, Arg arg, Status*) {
+    return std::round(arg);
+  }
+
+  template <typename T, typename Arg>
+  static constexpr enable_if_integer<T> Call(KernelContext*, Arg arg, Status*) {
+    return arg;
+  }
+};
+
+struct RoundChecked {
+  template <typename T, typename Arg>
+  static constexpr enable_if_integer<T> Call(KernelContext*, Arg arg, Status*) {
+    static_assert(std::is_same<T, Arg>::value, "");
+    return arg;
+  }
+
+  template <typename T, typename Arg>
+  static constexpr enable_if_floating_point<T> Call(KernelContext*, Arg arg, Status*) {
+    static_assert(std::is_same<T, Arg>::value, "");
```

Comment:
It's not worthwhile to have a _checked version of this function if there is no difference in runtime behavior between the two functions.

Reply:
I knew this comment was coming with PR as is. This is currently a placeholder for the cases where if the rounding options suggest to round to nearest integer and output will be of integral type, then overflow/underflow can occur.

Reply:
But should a `round` kernel ever change the type? IMO when rounding floats, the result can still be floats, regardless of the number of decimals to round

Reply:
I agree, but I was not sure if there was interest in mimicking C++ [lround](https://en.cppreference.com/w/cpp/numeric/math/round) variants. For now, it seems reasonable to only support floats and return result with same type.

Reply:
Ah, that might also be interesting, but I would indeed start with the type preserving round for now (there are also already enough complexities regarding round mode, whether we want ceil/floor/trunc etc)